README.md
# ts-quantities

Originally a fork of [gentooboontoo/js-quantities](https://github.com/gentooboontoo/js-quantities) rewritten in typescript.

## Installing ts-quantities

### Node

#### Install it in your project

```
npm install --save ts-quantities
```

#### Import it !

```typescript
import { Qty } from 'ts-quantities'
```

## Using ts-quantities

### Qty

ts-quantities heavily relies on `Qty` objects. To create `Qty` objects, use its constructor using `new`.

```typescript
qty = new Qty('23 ft');
```

`Qty` constructor accepts strings, numbers and `Qty` instances as
initializing values.

If scalars and their respective units are available programmatically, the
two argument signature may be useful:

```typescript
qty = new Qty(124, 'cm'); // => 1.24 meter
```

```typescript
qty = new Qty('1m'); // => 1 meter
qty = new Qty('m'); // =>  1 meter (scalar defaults to 1)

qty = new Qty('1 N*m');
qty = new Qty('1 N m'); // * is optional

qty = new Qty('1 m/s');

qty = new Qty('1 m^2/s^2');
qty = new Qty('1 m^2 s^-2'); // negative powers
qty = new Qty('1 m2 s-2'); // ^ is optional

qty = new Qty('1 m^2 kg^2 J^2/s^2 A');

qty = new Qty('1.5'); // unitless quantity
qty = new Qty(1.5); // number as initializing value

qty = new Qty('1 attoparsec/microfortnight');

qtyCopy = new Qty(qty); // quantity could be copied when used as
                        // initializing value
```

### Parsing a quantity manually

`Qty.parse` utility method is also provided to parse and create
quantities from strings. Unlike the constructor, it will return `null`
instead of throwing an error when parsing an invalid quantity.

```typescript
static parse(value: string): Qty

Qty.parse('1 m'); // => 1 meter
Qty.parse('foo'); // => null
```

### Available well-known kinds

```typescript
static getKinds(): string[]

Qty.getKinds(); // => Array of names of every well-known kind of units
```

### Available units of a particular kind

```typescript
static getUnits(kind?: string): string[]

Qty.getUnits('currency'); // => [ 'dollar', 'cents' ]
Qty.getUnits(); // All units, alphebetically sorted => [ 'acre','Ah','ampere','AMU','angstrom', ...]
```

### Alternative names of a unit

```typescript
static getAliases(unitName: string): string[]

Qty.getAliases('m'); // => [ 'm', 'meter', 'meters', 'metre', 'metres' ]
```

### Quantity compatibility, kind and various queries

```typescript
public isCompatible(other: Qty | string): boolean
qty1.isCompatible(qty2); // => true or false

public kind(): string
qty.kind(); // => 'length', 'area', etc...

public isUnitless(): boolean
qty.isUnitless(); // => true or false

public isBase(): boolean
qty.isBase(); // => true if quantity is represented with base units
```

### Conversion

```typescript
public toBase(): Qty
qty.toBase(); // converts to SI units (10 cm => 0.1 m) (new instance)

public toFloat(): number
qty.toFloat(); // returns scalar of unitless quantity
               // (otherwise throws error)

public to(Qty | string): Qty
qty.to('m'); // converts quantity to meter if compatible
             // or throws an error (new instance)
qty1.to(qty2); // converts quantity to same unit of qty2 if compatible
               // or throws an error (new instance)

public inverse(): Qty
qty.inverse(); // converts quantity to its inverse
               // ('100 m/s' => '0.01 s/m')
// Inverses can be used, but there is no special checking to
// rename the units
new Qty('10ohm').inverse(); // '0.1/ohm'
                       // (not '0.1S', although they are equivalent)
// however, the 'to' command will convert between inverses also
new Qty('10ohm').to('S'); // '0.1S'

public convertSingleUnit(baseUnit: string, targetUnit: string): Qty
// Converts the selected baseUnit into the target unit. Others units remain untouched.
new Qty('0.14 USD/kWh').convertSingleUnit('kWh', 'MWh'); // 140 USD/MWh
new Qty('4000 m2').convertSingleUnit('m', 'km'); // 0.004 km2
```

### Mass conversion

`Qty.swiftConverter()` is a fast way to efficiently convert large array of
Number values. It configures a function accepting a value or an array of Number
values to convert.

```typescript
static swiftConverter(srcUnits: string, dstUnits: string): (value: number | number[]) => number | number[]

const convert = Qty.swiftConverter('m/h', 'ft/s'); // Configures converter

// Converting single value
const converted = convert(2500); // => 2.278..
// Converting large array of values
const convertedArray = convert([2500, 5000, ...]); // => [2.278.., 4.556.., ...]
```

The main drawback of this conversion method is that it does not take care of
rounding issues.

### Comparison

```typescript
public eq(other: Qty): boolean
qty1.eq(qty2); // => true if both quantities are equal (1m == 100cm => true)

public same(other: Qty): boolean
qty1.same(qty2); // => true if both quantities are same (1m == 100cm => false)

public lt(other: Qty): boolean
qty1.lt(qty2); // => true if qty1 is stricty less than qty2

public lte(other: Qty): boolean
qty1.lte(qty2); // => true if qty1 is less than or equal to qty2

public gt(other: Qty): boolean
qty1.gt(qty2); // => true if qty1 is stricty greater than qty2

public gte(other: Qty): boolean
qty1.gte(qty2); // => true if qty1 is greater than or equal to qty2

public compareTo(other: Qty): number
qty1.compareTo(qty2); // => -1 if qty1 < qty2,
                      // => 0 if qty1 == qty2,
                      // => 1 if qty1 > qty2
```

### Operators

```typescript
public add(other: Qty | string | number): Qty // other should be unit compatible.
public sub(other: Qty | string | number): Qty // other should be unit compatible.
public mul(other: Qty | string | number): Qty
public div(other: Qty | string | number): Qty
```

### Rounding

`Qty#toPrec(precision)` : returns the nearest multiple of quantity passed as
precision.

```typescript
public toPrec(precQuantity: Qty | string | number): Qty

const qty = new Qty('5.17 ft');
qty.toPrec('ft'); // => 5 ft
qty.toPrec('0.5 ft'); // => 5 ft
qty.toPrec('0.25 ft'); // => 5.25 ft
qty.toPrec('0.1 ft'); // => 5.2 ft
qty.toPrec('0.05 ft'); // => 5.15 ft
qty.toPrec('0.01 ft'); // => 5.17 ft
qty.toPrec('0.00001 ft'); // => 5.17 ft
qty.toPrec('2 ft'); // => 6 ft
qty.toPrec('2'); // => 6 ft

const qty = new Qty('6.3782 m');
qty.toPrec('dm'); // => 6.4 m
qty.toPrec('cm'); // => 6.38 m
qty.toPrec('mm'); // => 6.378 m
qty.toPrec('5 cm'); // => 6.4 m
qty.toPrec('10 m'); // => 10 m
qty.toPrec(0.1); // => 6.3 m

const qty = new Qty('1.146 MPa');
qty.toPrec('0.1 bar'); // => 1.15 MPa
```

### Formatting quantities

`Qty#toString` returns a string using the canonical form of the quantity (that
is it could be seamlessly reparsed by `Qty`).

```typescript
public toString(toUnits?: string): string

var qty = new Qty('1.146 MPa');
qty.toString(); // => '1.146 MPa'
```

As a shorthand, units could be passed to `Qty#toString` and is equivalent to
successively call `Qty#to` then `Qty#toString`.

```typescript
const qty = new Qty('1.146 MPa');
qty.toString('bar'); // => '11.46 bar'
qty.to('bar').toString(); // => '11.46 bar'
```

`Qty#toString` could also be used with any method from `Qty` to make some sort
of formatting. For instance, one could use `Qty#toPrec` to fix the maximum
number of decimals:

```typescript
const qty = new Qty('1.146 MPa');
qty.toPrec(0.1).toString(); // => '1.1 MPa'
qty.to('bar').toPrec(0.1).toString(); // => '11.5 bar'
```

For advanced formatting needs as localization, specific rounding or any other
custom customization, quantities can be transformed into strings through
`Qty#format` according to optional target units and formatter. If target units
are specified, the quantity is converted into them before formatting.

Such a string is not intended to be reparsed to construct a new instance of
`Qty` (unlike output of `Qty#toString`).

If no formatter is specified, quantities are formatted according to default
ts-quantities' formatter and is equivalent to `Qty#toString`.

```typescript
type Formatter = (scalar: number, units: string) => string;
public format(formatter?: Formatter): string;
public format(targetUnits?: string, formatter?: Formatter): string;
```

```typescript
var qty = new Qty('1.1234 m');
qty.format(); // same units, default formatter => '1.234 m'
qty.format('cm'); // converted to 'cm', default formatter => '123.45 cm'
```

`Qty#format` could delegates formatting to a custom formatter if required. A
formatter is a callback function accepting scalar and units as parameters and
returning a formatted string representing the quantity.

```typescript
const configurableRoundingFormatter = (maxDecimals: number): Formatter => {
  return (scalar: number, units: string): string => {
    const pow = Math.pow(10, maxDecimals);
    const rounded = Math.round(scalar * pow) / pow;
    return rounded + ' ' + units;
  };
};

const qty = new Qty('1.1234 m');

// same units, custom formatter => '1.12 m'
qty.format(configurableRoundingFormatter(2));

// convert to 'cm', custom formatter => '123.4 cm'
qty.format('cm', configurableRoundingFormatter(1));
```

Custom formatter can be configured globally by setting `Qty.formatter`.

```typescript
static formatter: Formatter;

Qty.formatter = configurableRoundingFormatter(2);
const qty = new Qty('1.1234 m');
qty.format(); // same units, current default formatter => '1.12 m'
```

### Temperatures

Like ruby-units, ts-quantities makes a distinction between a temperature (which
technically is a property) and degrees of temperature (which temperatures are
measured in).

Temperature units (i.e., 'tempK') can be converted back and forth, and will take
into account the differences in the zero points of the various scales.
Differential temperature (e.g., '100 degC') units behave like most other units.

```typescript
new Qty('37 tempC').to('tempF') // => 98.6 tempF
```

ts-quantities will throw an error if you attempt to create a temperature unit
that would fall below absolute zero.

Unit math on temperatures is fairly limited.

```typescript
new Qty('100 tempC').add('10 degC')  // 110 tempC
new Qty('100 tempC').sub('10 degC')  // 90 tempC
new Qty('100 tempC').add('50 tempC') // throws error
new Qty('100 tempC').sub('50 tempC') // 50 degC
new Qty('50 tempC').sub('100 tempC') // -50 degC
new Qty('100 tempC').mul(scalar)     // 100*scalar tempC
new Qty('100 tempC').div(scalar)     // 100/scalar tempC
new Qty('100 tempC').mul(qty)        // throws error
new Qty('100 tempC').div(qty)        // throws error
new Qty('100 tempC*unit')            // throws error
new Qty('100 tempC/unit')            // throws error
new Qty('100 unit/tempC')            // throws error
new Qty('100 tempC').inverse()       // throws error
```

```typescript
new Qty('100 tempC').to('degC') // => 100 degC
```

This conversion references the 0 point on the scale of the temperature unit

```typescript
new Qty('100 degC').to('tempC') // => -173.15 tempC
```

These conversions are always interpreted as being relative to absolute zero.
Conversions are probably better done like this...

```typescript
new Qty('0 tempC').add('100 degC') // => 100 tempC
```

### Errors

Every error thrown by ts-quantities is an instance of `Qty.Error`.

```typescript
try {
  // code triggering an error inside ts-quantities
}
catch(error) {
  if(error instanceof Qty.Error) {
    // ...
  }
  else {
    // ...
  }
}
```

## Tests

Tests are implemented with Jasmine (https://github.com/pivotal/jasmine).

To execute specs through `jasmine-node`, launch:

```
npm install -g jasmine-node
npm run test
```

## Contribute

Feedback and contributions are welcomed.

Pull requests must pass tests and linting. Please make sure that `npm run test`
and `npm run lint` return no errors before submitting.

History.md
2.1.1 / 2020-02-18
------------------

* Update forked documentation to comply to typescript.

2.1.0 / 2020-02-18
------------------

* New member function `convertSingleUnit`.

2.0.0 / 2020-02-16
------------------

* Fork
* Rewrite in typescript

1.7.4 / 2019-06-26
------------------

* Add Imperial Gallons, and Barrels (US Beer, Imperial Beer, Oil)
* Add support for imperial version of fluid ounces and pints

1.7.3 / 2018-12-15
------------------

* Fix some inconsistent resulting units when multiplying or dividing
quantities (#94)

1.7.2 / 2018-03-31
------------------

* Remove `module` from package definition (Fix #91)

1.7.1 / 2018-03-25
------------------

* Add missing unit aliases

1.7.0 / 2017-08-23
------------------

* Fix JS allocation failure for unrealistically large exponents
* Add volt-ampere, volt-ampere-reactive and data mile definitions
* Add redshift alias

1.6.6 / 2017-02-08
------------------

* Report incompatible units in error message

1.6.5 / 2017-01-20
------------------

* Fix main entrypoint in bower.json

1.6.4 / 2017-01-02
------------------

* Fix infinite regex check

1.6.3 / 2016-09-22
------------------

* Add missing kinds
* Fix conversion from percentage to unitless quantity
* Fix capacitance definition
* Exclude `farad` from base units
* Rename `mass concentration` to `density`

1.6.2 / 2016-04-13
------------------

* Accept blank string as initialization value

1.6.1 / 2016-03-27
------------------

* Fix definition of square foot
* Add tablespoon `tbsp` alias
* Add `Qty.version` property

1.6.0 / 2015-12-26
------------------

* Add `Qty.getUnits` to return available units of a well-known kind
* Add `Qty.getAliases` to return every alias of a specific unit
* Allow to initialize a quantity with scalar and units as separate arguments
* Rename `memory` kind to `information`
* Add `information_rate` kind
* Accept Wh and Ah as units
* Fix hang when using water height pressure units
* Add plural for fluid ounce
* Fix `amu` and `dalton` definitions
* Add `tb` as tablespoon alias
* Add plural for `information` units
* Minor fixes or improvements

1.5.0 / 2014-12-08
------------------

* Add `Qty.getKinds` returning known kinds of units
* Add µ symbol as micro prefix alias
* Add Ω symbol as ohm unit alias
* Minor internal improvements and fixes

1.4.2 / 2014-09-09
------------------

* Fix plural for radian and add missing ones for time units
* Add "gon" international standard symbol as gradian alias
* Fix units of force
* Allow whitespaces between sign and scalar and do not accept sign
  without scalar

1.4.1 / 2014-05-14
------------------

* Use a little more robust to test string type and factorize it

1.4.0 / 2014-04-10
------------------

* Directly convert array of values when using swiftConverter
* Add support for bushel units

1.3.0 / 2014-03-05
------------------

* Add Qty#format and accept custom formatters
* Allow to call Qty() without new to create Qty instances (Qty could be used
  both as a constructor or as a factory)
* Qty#toString only supports to be passed output units as single parameter.
  Former parameters are now deprecated but still supported to not introduce
  a breaking change
* Add mc as alternate definition for prefix "micro"
* Throw error with mmm as unit
* Add rounding optimization

1.2.0 / 2013-12-17
------------------

* Throw QtyError instead of plain string
* Cache conversion results from Qty#to instead of Qty#toString
* Fix point and pica unit definitions
* Fix error when initializing a quantity with an empty string

1.1.2 / 2013-11-04
------------------

* Fix rounding issue when converting 1 cm3 to mm3
* Do some code cleaning (it should not break public API)

1.1.1 / 2013-10-01
------------------

* Fix Qty#toPrec() returning wrong result with some precision

1.1.0 / 2013-09-20
------------------

* Add array converting method
* Major speedup by means of some optimizations and refactoring

1.0.0 / 2013-07-30
------------------

* First stable version


src/quantities/kind.ts
import { Qty } from './constructor.js';
import { uniq } from './utils.js';
import { RegularObject } from './types.js';

const KINDS: RegularObject<string> = {
    '-312078': 'elastance',
    '-312058': 'resistance',
    '-312038': 'inductance',
    '-152058': 'potential',
    '-152040': 'magnetism',
    '-152038': 'magnetism',
    '-7997': 'specific_volume',
    '-79': 'snap',
    '-59': 'jolt',
    '-39': 'acceleration',
    '-38': 'radiation',
    '-20': 'frequency',
    '-19': 'speed',
    '-18': 'viscosity',
    '-17': 'volumetric_flow',
    '-1': 'wavenumber',
    '0': 'unitless',
    '1': 'length',
    '2': 'area',
    '3': 'volume',
    '20': 'time',
    '400': 'temperature',
    '7941': 'yank',
    '7942': 'power',
    '7959': 'pressure',
    '7961': 'force',
    '7962': 'energy',
    '7979': 'viscosity',
    '7981': 'momentum',
    '7982': 'angular_momentum',
    '7997': 'density',
    '7998': 'area_density',
    '8000': 'mass',
    '152020': 'radiation_exposure',
    '159999': 'magnetism',
    '160000': 'current',
    '160020': 'charge',
    '312058': 'conductance',
    '312078': 'capacitance',
    '3199980': 'activity',
    '3199997': 'molar_concentration',
    '3200000': 'substance',
    '63999998': 'illuminance',
    '64000000': 'luminous_power',
    '1280000000': 'currency',
    '25599999980': 'information_rate',
    '25600000000': 'information',
    '511999999980': 'angular_velocity',
    '512000000000': 'angle'
};

/**
 * Returns the list of available well-known kinds of units, e.g.
 * "radiation" or "length".
 *
 * @returns {string[]} names of kinds of units
 */
export function getKinds(): string[] {
    return uniq(
        Object.keys(KINDS).map(knownSignature => {
            return KINDS[knownSignature];
        })
    );
}

export function kind(this: Qty): string {
    return KINDS[this.signature.toString()];
}
